lib/ipam/ipam.go

```diff
@@ -548,7 +549,44 @@ func (s *blockAssignState) findOrClaimBlock(ctx context.Context, minFreeIps int)
 	return nil, false, errors.New("failed to find or claim a block")
 }
 
-func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, attrs map[string]string, requestedPools []net.IPNet, version int, host string, maxNumBlocks int, rsvdAttr *HostReservedAttr) ([]net.IPNet, error) {
+type IPAMAssignmentInfo struct {
+	NumRequested       int
+	NumAssigned        int
+	IPVersion          int
+	NumBlocksOwned     int
+	MaxNumBlocks       int
+	ExhaustedPools     []string
+	StrictAffinity     bool
+	NoFreeAffineBlocks bool
+	HostReservedAttr   *HostReservedAttr
+}
+
+func (i *IPAMAssignmentInfo) String() string {
+	if i == nil {
+		return ""
+	}
+
+	ret := fmt.Sprintf("Assigned %v out of %v requested IPv%v addresses", i.NumAssigned, i.NumRequested, i.IPVersion)
+
+	switch {
+	case i.NoFreeAffineBlocks && i.StrictAffinity:
+		ret += "; No more free affine blocks and strict affinity enabled"
+	case i.NumBlocksOwned >= i.MaxNumBlocks:
+		ret += "; IPAM block limit reached"
+	case len(i.ExhaustedPools) > 0:
+		ret += "; No IPs available in pools"
+	}
+	if i.NumAssigned != i.NumRequested {
+		ret += fmt.Sprintf("; no free affine blocks: %v, strict affinity: %v, assigned blocks: %v, block limit: %v, exhausted IP pools: %v", i.NoFreeAffineBlocks, i.StrictAffinity, i.NumBlocksOwned, i.MaxNumBlocks, i.ExhaustedPools)
```

Comment:
I wonder if this is too much information to display on every failed request. For example, if we hit the block limit for a node, do we need to say anything about exhausted IP pools? Or can we limit the message to just the issue that was actually encountered?

Reply:
I had initially thought this way, but after some discussion with @mgleung I ended up including everything. I can see advantages and disadvantages to both, I don't know... Maybe we could have a `DumpValues()` method which does this, and defer the choice of calling either `String()` or `DumpValues()` (not married to this name, though), or both?